### PR TITLE
RavenDB-4329 indexing base

### PR DESCRIPTION
**Pull request by ppekrol**

### CODE REVIEW

**Comment by ppekrol** [comment]
Currently I'm waiting in IndexStore.CreateIndex for OpenIndexes to complete, I think we can do better here e.g. book the indexId and use fake instance that would be returning exception that index is loading if someone is trying to Query that index. If it loads then swap to that instance. Thoughts?

**Comment by ayende** [comment]
Why do we need to do this? So we can speed up the server startup time?

File: src/Raven.Server/Documents/Indexes/IndexStore.cs [hunk 1]
```diff
@@ -365,8 +369,23 @@ private void OpenIndexes()
...
-                        _indexes.Add(index);
+                        try
+                        {
+                            var index = Index.Open(indexId, _documentDatabase);
+                            _indexes.Add(index);
+                        }
+                        catch (Exception e)
+                        {
+                            // TODO arek: I think we can ignore auto indexes here, however for static ones try to retrieve names
+                            var fakeIndex = new FaultyInMemoryIndex(indexId);
```
Comment (ayende) [inline src/Raven.Server/Documents/Indexes/IndexStore.cs]: Why are we passing just the id? Why not the name as well?
Don't we have a name here?

**Comment by ayende** [comment]
I don't think that you need to do it in this fashion, see my changes on top of your code.
We start opening the indexes in a background thread, continue doing other initialization, and then wait for the index initialization to complete.
That is much easier for us to reason about.

File: src/Raven.Server/Documents/Indexes/IndexStore.cs (hunk 1 above)
Comment (arekpalinski) [inline src/Raven.Server/Documents/Indexes/IndexStore.cs]: Yes. We don't have a name here. Is there any reason to still keep this code (creation of `FaultyInMemoryIndex`) after your changes? We won't allow to open the database anyway if any exception occurs here.

Comment (ayende) [inline src/Raven.Server/Documents/Indexes/IndexStore.cs]: Yes, we need it.
Note that a failure of a single index (or all indexes) is fine, and will result in the faulty index showing up there.
We try to load each index, and we won't fail the whole open process if we can't open an index.
https://github.com/ayende/ravendb/blob/v4.0/src/Raven.Server/Documents/Indexes/IndexStore.cs#L378

Only if the _entire indexing process is broken_ (the indexing directory is not found, for example, or inaccessible), or something like this. Will we abort it.

**Pull request merged**